File: 1_stack/括号/F - Parenthesis Checking.py
"""验证合法的括号序列

将左括号和右括号分别看作是 1 和 -1,那么有效的括号序列条件为
**min(preSum[i]) >= 0 and preSum[-1] == 0**

线段树维护区间前缀和的最小值
"""

from itertools import accumulate
import sys
from typing import List, Union

sys.setrecursionlimit(int(1e6))
input = lambda: sys.stdin.readline().rstrip("\r\n")
MOD = 998244353
INF = int(4e18)


class MinSegmentTree2:
    """RMQ 最小值(区间和叠加) 线段树

    基元为0 超出范围返回0
    注意根节点从1开始,tree本身为[1,n]
    因为是叠加 所以不需要isLazy数组
    """

    __slots__ = "_n", "_tree", "_lazy"

    def __init__(self, nOrNums: Union[int, List[int]]):
        self._n = nOrNums if isinstance(nOrNums, int) else len(nOrNums)
        self._tree = [0] * (4 * self._n)
        self._lazy = [0] * (4 * self._n)
        if isinstance(nOrNums, list):
            self._build(1, 1, self._n, nOrNums)

    def queryAll(self) -> int:
        return self._tree[1]

    def query(self, left: int, right: int) -> int:
        """闭区间[left,right]的最小值"""
        if left < 1:
            left = 1
        if right > self._n:
            right = self._n
        if left > right:
            return 0  # !超出范围返回0
        return self._query(1, left, right, 1, self._n)

    def add(self, left: int, right: int, delta: int) -> None:
        """闭区间[left,right]区间值加上delta"""
        if left < 1:
            left = 1
        if right > self._n:
            right = self._n
        if left > right:
            return
        self._add(1, left, right, 1, self._n, delta)

    def _build(self, rt: int, l: int, r: int, nums: List[int]) -> None:
        """传了nums时,用于初始化线段树"""
        if l == r:
            self._tree[rt] = nums[l - 1]
            return

        mid = (l + r) // 2
        self._build(rt * 2, l, mid, nums)
        self._build(rt * 2 + 1, mid + 1, r, nums)
        self._push_up(rt)

    def _add(self, rt: int, L: int, R: int, l: int, r: int, delta: int) -> None:
        """L,R表示需要add的范围,l,r表示当前节点的范围"""
        if L <= l and r <= R:
            self._lazy[rt] += delta
            self._tree[rt] += delta
            return

        mid = (l + r) // 2
        self._push_down(rt, l, r, mid)
        if L <= mid:
            self._add(rt * 2, L, R, l, mid, delta)
        if mid < R:
            self._add(rt * 2 + 1, L, R, mid + 1, r, delta)
        self._push_up(rt)

    def _query(self, rt: int, L: int, R: int, l: int, r: int) -> int:
        """L,R表示需要query的范围,l,r表示当前节点的范围"""
        if L <= l and r <= R:
            return self._tree[rt]

        # 传递懒标记
        mid = (l + r) // 2
        self._push_down(rt, l, r, mid)
        res = INF  # !默认值为INF
        if L <= mid:
            cand = self._query(rt * 2, L, R, l, mid)
            if cand < res:
                res = cand
        if mid < R:
            cand = self._query(rt * 2 + 1, L, R, mid + 1, r)
            if cand < res:
                res = cand
        return res

    def _push_up(self, rt: int) -> None:
        self._tree[rt] = self._tree[rt * 2]
        if self._tree[rt * 2 + 1] < self._tree[rt]:
            self._tree[rt] = self._tree[rt * 2 + 1]

    def _push_down(self, rt: int, l: int, r: int, mid: int) -> None:
        if self._lazy[rt]:
            value = self._lazy[rt]
            self._lazy[rt * 2] += value
            self._lazy[rt * 2 + 1] += value

            self._tree[rt * 2] += value
            self._tree[rt * 2 + 1] += value

            self._lazy[rt] = 0


if __name__ == "__main__":
    n, q = map(int, input().split())
    s = input()  # 括号序列
    nums = [1 if c == "(" else -1 for c in s]
    preSum = MinSegmentTree2(list(accumulate(nums)))  # !线段树维护前缀和

    for _ in range(q):
        kind, left, right = map(int, input().split())  # 1<=left<=right<=n

        if kind == 1:  # !交换s[left]和s[right]
            if nums[left - 1] == nums[right - 1]:
                continue
            if nums[left - 1] == 1 and nums[right - 1] == -1:
                preSum.add(left, right - 1, -2)
            elif nums[left - 1] == -1 and nums[right - 1] == 1:
                preSum.add(left, right - 1, 2)
            nums[left - 1], nums[right - 1] = nums[right - 1], nums[left - 1]
        else:  # !检查s[left:right+1]是否为合法的括号序列
            pre = preSum.query(left - 1, left - 1) if left > 1 else 0
            min_ = preSum.query(left, right) - pre
            last = preSum.query(right, right) - pre
            print("Yes" if (min_ >= 0 and last == 0) else "No")

File: 1_stack/括号/F - Parenthesis Checking.go
package main

import (
	"bufio"
	"fmt"
	"math/bits"
	"os"
	"strings"
)

// F - Parenthesis Checking
// https://atcoder.jp/contests/abc223/tasks/abc223_f
// 给出一个括号串s和Q 次以下两种操作：
// 1 l r，代表交换第 l 和第 r 个位置上的字符
// 2 l r，判断区间 [l,r] 子串是否是合法括号序列
//
// 验证合法的括号序列
// 将左括号和右括号分别看作是 1 和 -1,那么有效的括号序列条件为
// - 总共的左括号和右括号数量相等 => 和为0
// - 任意前缀的左括号数量 ≥ 右括号数量 => 任意前缀和>=0，即前缀和的最小值>=0
// 线段树维护前缀和的最小值即可.
func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, q int
	fmt.Fscan(in, &n, &q)
	var s string
	fmt.Fscan(in, &s)

	P := NewParenthesisChecking(n, func(i int) bool {
		if s[i] == '(' {
			return true
		}
		return false
	})

	for i := 0; i < q; i++ {
		var kind, start, end int
		fmt.Fscan(in, &kind, &start, &end)
		start--

		if kind == 1 {
			P.Swap(start, end-1)
		} else {
			if P.Query(start, end) {
				fmt.Fprintln(out, "Yes")
			} else {
				fmt.Fprintln(out, "No")
			}
		}
	}
}

type ParenthesisChecking struct {
	brackets []bool
	minSeg   *LazySegTree
}

// ( => true
// ) => false
func NewParenthesisChecking(n int, f func(i int) bool) *ParenthesisChecking {
	brackets := make([]bool, n)
	for i := 0; i < n; i++ {
		brackets[i] = f(i)
	}
	preSum := make([]int32, n+1)
	for i := 0; i < n; i++ {
		if brackets[i] {
			preSum[i+1] = preSum[i] + 1
		} else {
			preSum[i+1] = preSum[i] - 1
		}
	}
	minSeg := NewLazySegTreeFrom(preSum)
	return &ParenthesisChecking{brackets: brackets, minSeg: minSeg}
}

func (pc *ParenthesisChecking) Query(start, end int) bool {
	seg := pc.minSeg
	min_ := seg.Query(start, end+1)
	preSum1 := seg.Get(start)
	min_ -= preSum1
	if min_ < 0 { // 任意前缀和>=0
		return false
	}
	preSum2 := seg.Get(end)
	return preSum2 == preSum1 // 总共的左括号和右括号数量相等
}

func (pc *ParenthesisChecking) Swap(i, j int) {
	if pc.brackets[i] == pc.brackets[j] {
		return
	}
	if pc.brackets[i] && !pc.brackets[j] { // ( => )
		pc.minSeg.Update(i+1, j+1, -2)
	}
	if !pc.brackets[i] && pc.brackets[j] { // ) => (
		pc.minSeg.Update(i+1, j+1, 2)
	}
	pc.brackets[i], pc.brackets[j] = pc.brackets[j], pc.brackets[i]
}

const INF32 = 1e9 + 10

// RangeAddRangeMin

type E = int32
type Id = int32

func (*LazySegTree) e() E                   { return INF32 }
func (*LazySegTree) id() Id                 { return 0 }
func (*LazySegTree) op(e1, e2 E) E          { return min32(e1, e2) }
func (*LazySegTree) mapping(f Id, g E) E    { return f + g }
func (*LazySegTree) composition(f, g Id) Id { return f + g }

type LazySegTree struct {
	n    int
	size int
	log  int
	data []E
	lazy []Id
}

func NewLazySegTree(n int, f func(int) E) *LazySegTree {
	tree := &LazySegTree{}
	tree.n = n
	tree.log = int(bits.Len(uint(n - 1)))
	tree.size = 1 << tree.log
	tree.data = make([]E, tree.size<<1)
	tree.lazy = make([]Id, tree.size)
	for i := range tree.data {
		tree.data[i] = tree.e()
	}
	for i := range tree.lazy {
		tree.lazy[i] = tree.id()
	}
	for i := 0; i < n; i++ {
		tree.data[tree.size+i] = f(i)
	}
	for i := tree.size - 1; i >= 1; i-- {
		tree.pushUp(i)
	}
	return tree
}

func NewLazySegTreeFrom(leaves []E) *LazySegTree {
	tree := &LazySegTree{}
	n := len(leaves)
	tree.n = n
	tree.log = int(bits.Len(uint(n - 1)))
	tree.size = 1 << tree.log
	tree.data = make([]E, tree.size<<1)
	tree.lazy = make([]Id, tree.size)
	for i := range tree.data {
		tree.data[i] = tree.e()
	}
	for i := range tree.lazy {
		tree.lazy[i] = tree.id()
	}
	for i := 0; i < n; i++ {
		tree.data[tree.size+i] = leaves[i]
	}
	for i := tree.size - 1; i >= 1; i-- {
		tree.pushUp(i)
	}
	return tree
}

// 查询切片[left:right]的值
//
//	0<=left<=right<=len(tree.data)
func (tree *LazySegTree) Query(left, right int) E {
	if left < 0 {
		left = 0
	}
	if right > tree.n {
		right = tree.n
	}
	if left >= right {
		return tree.e()
	}
	left += tree.size
	right += tree.size
	for i := tree.log; i >= 1; i-- {
		if ((left >> i) << i) != left {
			tree.pushDown(left >> i)
		}
		if ((right >> i) << i) != right {
			tree.pushDown((right - 1) >> i)
		}
	}
	sml, smr := tree.e(), tree.e()
	for left < right {
		if left&1 != 0 {
			sml = tree.op(sml, tree.data[left])
			left++
		}
		if right&1 != 0 {
			right--
			smr = tree.op(tree.data[right], smr)
		}
		left >>= 1
		right >>= 1
	}
	return tree.op(sml, smr)
}
func (tree *LazySegTree) QueryAll() E {
	return tree.data[1]
}
func (tree *LazySegTree) GetAll() []E {
	res := make([]E, tree.n)
	copy(res, tree.data[tree.size:tree.size+tree.n])
	return res
}

// 更新切片[left:right]的值
//
//	0<=left<=right<=len(tree.data)
func (tree *LazySegTree) Update(left, right int, f Id) {
	if left < 0 {
		left = 0
	}
	if right > tree.n {
		right = tree.n
	}
	if left >= right {
		return
	}
	left += tree.size
	right += tree.size
	for i := tree.log; i >= 1; i-- {
		if ((left >> i) << i) != left {
			tree.pushDown(left >> i)
		}
		if ((right >> i) << i) != right {
			tree.pushDown((right - 1) >> i)
		}
	}
	l2, r2 := left, right
	for left < right {
		if left&1 != 0 {
			tree.propagate(left, f)
			left++
		}
		if right&1 != 0 {
			right--
			tree.propagate(right, f)
		}
		left >>= 1
		right >>= 1
	}
	left = l2
	right = r2
	for i := 1; i <= tree.log; i++ {
		if ((left >> i) << i) != left {
			tree.pushUp(left >> i)
		}
		if ((right >> i) << i) != right {
			tree.pushUp((right - 1) >> i)
		}
	}
}

// 二分查询最小的 left 使得切片 [left:right] 内的值满足 predicate
func (tree *LazySegTree) MinLeft(right int, predicate func(data E) bool) int {
	if right == 0 {
		return 0
	}
	right += tree.size
	for i := tree.log; i >= 1; i-- {
		tree.pushDown((right - 1) >> i)
	}
	res := tree.e()
	for {
		right--
		for right > 1 && right&1 != 0 {
			right >>= 1
		}
		if !predicate(tree.op(tree.data[right], res)) {
			for right < tree.size {
				tree.pushDown(right)
				right = right<<1 | 1
				if tmp := tree.op(tree.data[right], res); predicate(tmp) {
					res = tmp
					right--
				}
			}
			return right + 1 - tree.size
		}
		res = tree.op(tree.data[right], res)
		if (right & -right) == right {
			break
		}
	}
	return 0
}

// 二分查询最大的 right 使得切片 [left:right] 内的值满足 predicate
func (tree *LazySegTree) MaxRight(left int, predicate func(data E) bool) int {
	if left == tree.n {
		return tree.n
	}
	left += tree.size
	for i := tree.log; i >= 1; i-- {
		tree.pushDown(left >> i)
	}
	res := tree.e()
	for {
		for left&1 == 0 {
			left >>= 1
		}
		if !predicate(tree.op(res, tree.data[left])) {
			for left < tree.size {
				tree.pushDown(left)
				left <<= 1
				if tmp := tree.op(res, tree.data[left]); predicate(tmp) {
					res = tmp
					left++
				}
			}
			return left - tree.size
		}
		res = tree.op(res, tree.data[left])
		left++
		if (left & -left) == left {
			break
		}
	}
	return tree.n
}

// 单点查询(不需要 pushUp/op 操作时使用)
func (tree *LazySegTree) Get(index int) E {
	index += tree.size
	for i := tree.log; i >= 1; i-- {
		tree.pushDown(index >> i)
	}
	return tree.data[index]
}

// 单点赋值
func (tree *LazySegTree) Set(index int, e E) {
	index += tree.size
	for i := tree.log; i >= 1; i-- {
		tree.pushDown(index >> i)
	}
	tree.data[index] = e
	for i := 1; i <= tree.log; i++ {
		tree.pushUp(index >> i)
	}
}

func (tree *LazySegTree) pushUp(root int) {
	tree.data[root] = tree.op(tree.data[root<<1], tree.data[root<<1|1])
}
func (tree *LazySegTree) pushDown(root int) {
	if tree.lazy[root] != tree.id() {
		tree.propagate(root<<1, tree.lazy[root])
		tree.propagate(root<<1|1, tree.lazy[root])
		tree.lazy[root] = tree.id()
	}
}
func (tree *LazySegTree) propagate(root int, f Id) {
	tree.data[root] = tree.mapping(f, tree.data[root])
	// !叶子结点不需要更新lazy
	if root < tree.size {
		tree.lazy[root] = tree.composition(f, tree.lazy[root])
	}
}

func (tree *LazySegTree) String() string {
	var sb []string
	sb = append(sb, "[")
	for i := 0; i < tree.n; i++ {
		if i != 0 {
			sb = append(sb, ", ")
		}
		sb = append(sb, fmt.Sprintf("%v", tree.Get(i)))
	}
	sb = append(sb, "]")
	return strings.Join(sb, "")
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
func max(a, b int) int {
	if a < b {
		return b
	}
	return a
}

func min32(a, b int32) int32 {
	if a < b {
		return a
	}
	return b
}
